# MSIS 2629 Spring 2019-Gun Violence in United States Project

## Team Members:

Xinran Li

Xin Zou

Chia-cheng Lin

Jui-Chuan Ma

## Dateset Exploration Roadmap

We started by [compiling all the datasets](https://docs.google.com/document/d/18wLsW_Fxxjvqqny4Dy7Dy-pm7umAD9WdH6EGEO0ggjo/edit) from the Vox article regarding gun violence in America to see the different kinds of datasets that are out there. There are only a handful of datasets that are easy to download so we started to explore those datasets.

The first one we found was the [gun violence record from 2013 - 2018 dataset](https://www.kaggle.com/jameslko/gun-violence-data) from Kaggle but quickly realized that a lot of variables are not clearly defined. We tried to capture this data directly from the source ([Gun Violence Archive](https://www.gunviolencearchive.org/)) but was not able to replicate the same data due to website limitations. The website only allowed us to download a fraction of the data at once and have no way to scrape the rest of the data.

The second one is a dataset that we decided to go with. This dataset contains information about [gun-deaths in the US from 2012 to 2014](https://www.kaggle.com/hakabuk/gun-deaths-in-the-us). This dataset is very clean and contains unique information such as victims'education and shooter's intent. There is also the police variable that indicates whether the police was involved, this will be interesting to examine comparing with the same variable in the previous dataset. The limitation of this dataset is that the time span is only three years. It will be hard to establish a trend with only three years of data. 

The last one is another dataset from Kaggle. This one is [mass shootings from 1966 to 2017](https://www.kaggle.com/zusmani/us-mass-shootings-last-50-years). We made an assumption that “mass shootings” is defined by both the intent of the shooter and the location of the shooter; the event usually takes place at a public/crowded place and/or the shooter has an intent to harm multiple people.  
This dataset is relatively clean, but we still made some modifications on some columns that will enable us to quickly generate visualizations. The limitation of this dataset is that three useful variables (Age,Employeed situation, Employed at) have more than 50% N/A varibles, it limits us to explore some questions related to these variables. Finally, because the dataset still has many useful information such as mass shooting location, target people and so on, we decided to go with this dataset even though it has some limitations.


## Dataset Discription

### gun-deaths in the US from 2012 to 2014 description

year: The year in which the fatality occurred.  

month: The month in which the fatality occurred.

intent : The intent of the perpetrator of the crime. This can be Suicide, Accidental, NA, Homicide, or Undetermined 

police : Whether a police officer was involved with the shooting. Either 0 (false) or 1 (true)

sex : The gender of the victim. Either M or F.

age : The age of the victim.

race : The race of the victim. Either Asian/Pacific Islander, Native American/Native Alaskan, Black, Hispanic, or White.  

hispanic: A code indicating the Hispanic origin of the victim.

place : Where the shooting took place

education : Educational status of the victim. Can be one of the following: 1: Less than High School 2: Graduated from High School or equivalent 3: Some College 4: At least graduated from College 5: Not available

### mass shootings from 1966 to 2017 Discription

Location: Location where the shooting took place

Date: Date of the shooting

Open/Close Location: Inside the building or open space

Target: Possible target audience or company  

Mental Health Issues: Whether the shooter has a mental health issue

Cause: The reason causes the shooting. Terrorism, Hate Crime, Fun (for no obvious reason etc.)  

Age: Shooter's age  

Employeed (Y/N): Whether the shooter has employeed or not
 
Employed at: Shooter's employeed location

Fatalities: Total number of people deceased

Injured: Total number of people injured

Policeman Killed: Total number of policeman killed in that event

Race: Race of the shooter

Gender: Gender of the shooter

Latitude/Longtitude: Location of the event

## Clean Date Process

In [6]:
import pandas as pd
import numpy as np
df_death = pd.read_csv("guns.csv")
df_mass = pd.read_csv("Mass.csv", encoding = 'ISO-8859-1')

### Part1-Below is the code on how we clean the data for gun-deaths in the US from 2012 to 2014 

In [8]:
pd.options.display.max_columns = 999

In [9]:
df_death.head()

Unnamed: 0  year  month   intent  police sex   age                    race  \
0           1  2012      1  Suicide       0   M  34.0  Asian/Pacific Islander   
1           2  2012      1  Suicide       0   F  21.0                   White   
2           3  2012      1  Suicide       0   M  60.0                   White   
3           4  2012      2  Suicide       0   M  64.0                   White   
4           5  2012      2  Suicide       0   M  31.0                   White   

   hispanic            place  education  
0       100             Home        4.0  
1       100           Street        3.0  
2       100  Other specified        4.0  
3       100             Home        4.0  
4       100  Other specified        2.0

In [10]:
df_death.isna().sum()

Unnamed: 0       0
year             0
month            0
intent           1
police           0
sex              0
age             18
race             0
hispanic         0
place         1384
education       53
dtype: int64

In [11]:
len(df_death)

100798

This dataset is very clean, and we have 100798 rows of data. We decided to drop the first column because it is useless and all rows that have NA because they only take up a small portion of our data

In [12]:
df2 = df_death[df_death.isna().sum(1)==0]

In [13]:
df2 = df2.drop(columns={"Unnamed: 0"})

In [14]:
df2.hispanic.unique()

array([100, 281, 211, 998, 261, 210, 222, 282, 260, 270, 231, 237, 200,
       223, 226, 275, 250, 234, 280, 227, 224, 286, 233, 271, 220, 225,
       235, 242, 212, 221, 239, 299, 232, 291, 217, 252, 209, 238, 218])

We already have a race column so "Hispanic" is not needed

In [15]:
df2 = df2.drop(columns={"hispanic"})

In [16]:
df2.race.unique()

array(['Asian/Pacific Islander', 'White',
       'Native American/Native Alaskan', 'Black', 'Hispanic'],
      dtype=object)

In [17]:
df2.place.unique()

array(['Home', 'Street', 'Other specified', 'Other unspecified',
       'Trade/service area', 'Farm', 'Industrial/construction',
       'School/instiution', 'Sports', 'Residential institution'],
      dtype=object)

In [18]:
df2.intent.unique()

array(['Suicide', 'Undetermined', 'Accidental', 'Homicide'], dtype=object)

In [19]:
df2.sex.unique()

array(['M', 'F'], dtype=object)

In [20]:
df2.education.unique()

array([4., 3., 2., 1., 5.])

Education is a numeric value, so we decided to put the meaning of the number in the data

1: Less than High School 

2: Graduated from High School or equivalent 

3: Some College 

4: At least graduated from College 

5: Not available

In [21]:
df2["Education"] = df2["education"].apply(lambda a: "Less than High School" if a == 1.0 else "Graduated from High School or equivalent" if a ==2.0\
                      else "Some College" if a == 3.0 else "At least graduated from College" if a == 4.0 else "Not available")

In [22]:
df2 = df2.drop(columns="education")

In [23]:
df2.head()

year  month   intent  police sex   age                    race  \
0  2012      1  Suicide       0   M  34.0  Asian/Pacific Islander   
1  2012      1  Suicide       0   F  21.0                   White   
2  2012      1  Suicide       0   M  60.0                   White   
3  2012      2  Suicide       0   M  64.0                   White   
4  2012      2  Suicide       0   M  31.0                   White   

             place                                 Education  
0             Home           At least graduated from College  
1           Street                              Some College  
2  Other specified           At least graduated from College  
3             Home           At least graduated from College  
4  Other specified  Graduated from High School or equivalent

Double check the cleaning result

In [24]:
df2.isna().sum()

year         0
month        0
intent       0
police       0
sex          0
age          0
race         0
place        0
Education    0
dtype: int64

##### Ideas that we can explore with the gun-death dataset

- Intent of the the shooter 
- Events where the police was involved
- Education of the shooter
- Age of the shooter
- Time variable compare with all of the variables mentioned above


### Part 2-Below is the code on how we clean the data for gun-deaths in the US from mass shootings from 1966 to 2017

We were curious to see what defines as mass shootings. We investigated the total victim count but the minimum record is as low as 3 victims.

In [25]:
df_mass['Total victims'].min()

3

Even though the data contains over 50 years of data, the dataset is relatively small with only 323 rows of data. 

In [26]:
len(df_mass)

323

We believe that this is acceptable given that the nature of mass shooting events does not happen often. There are some interesting variables that we can explore with visualization such as gender, race, number of victims, and how many policeman were killed. We also have time data that we can possibly draw trends from.

In [27]:
df_mass.isna().sum()

S#                        0
Title                     0
Location                 45
Date                      0
Incident Area            62
Open/Close Location      28
Target                    5
Cause                    77
Summary                   0
Fatalities                0
Injured                   0
Total victims             0
Policeman Killed          6
Age                     144
Employeed (Y/N)         256
Employed at             282
Mental Health Issues      0
Race                      2
Gender                    0
Latitude                 20
Longitude                20
dtype: int64

We decided to drop columns that have a lot of NAs, as well as columns that are not useful in our upcoming visualization

For the "Location", we unified them into just State Code for the better statistic.

In [28]:
us_state_abbrev = {
    'Alabama': 'AL',
    'Alaska': 'AK',
    'Arizona': 'AZ',
    'Arkansas': 'AR',
    'California': 'CA',
    'Colorado': 'CO',
    'Connecticut': 'CT',
    'Delaware': 'DE',
    'Florida': 'FL',
    'Georgia': 'GA',
    'Hawaii': 'HI',
    'Idaho': 'ID',
    'Illinois': 'IL',
    'Indiana': 'IN',
    'Iowa': 'IA',
    'Kansas': 'KS',
    'Kentucky': 'KY',
    'Louisiana': 'LA',
    'Maine': 'ME',
    'Maryland': 'MD',
    'Massachusetts': 'MA',
    'Michigan': 'MI',
    'Minnesota': 'MN',
    'Mississippi': 'MS',
    'Missouri': 'MO',
    'Montana': 'MT',
    'Nebraska': 'NE',
    'Nevada': 'NV',
    'New Hampshire': 'NH',
    'New Jersey': 'NJ',
    'New Mexico': 'NM',
    'New York': 'NY',
    'North Carolina': 'NC',
    'North Dakota': 'ND',
    'Ohio': 'OH',
    'Oklahoma': 'OK',
    'Oregon': 'OR',
    'Pennsylvania': 'PA',
    'Rhode Island': 'RI',
    'South Carolina': 'SC',
    'South Dakota': 'SD',
    'Tennessee': 'TN',
    'Texas': 'TX',
    'Utah': 'UT',
    'Vermont': 'VT',
    'Virginia': 'VA',
    'Washington': 'WA',
    'West Virginia': 'WV',
    'Wisconsin': 'WI',
    'Wyoming': 'WY',
}

In [29]:
df1 = df_mass.drop(columns = ['Age','Employeed (Y/N)','Employed at','Incident Area','S#','Title','Summary','Mental Health Issues'])

In [30]:
df1["State"] = df1['Location'].apply(
    lambda location: location.split(',')[-1].strip() if isinstance(location, str) else location
)

In [31]:
df1['State'] = df1["State"].apply(
    lambda state: us_state_abbrev[state] if state in us_state_abbrev else state
)

In [32]:
df1 = df1.drop(columns="Location")

In [33]:
df1.head(10)

Date Open/Close Location     Target          Cause  Fatalities  \
0   11/5/17               Close     random        unknown          26   
1   11/1/17                Open     random        unknown           3   
2  10/18/17               Close  coworkers        unknown           3   
3   10/1/17                Open     random        unknown          59   
4   6/14/17               Close  coworkers            NaN           3   
5    6/7/17               Close  coworkers      terrorism           3   
6    6/5/17               Close  coworkers  unemployement           5   
7   5/12/17               Close  coworkers            NaN           3   
8   4/18/17                Open     random         racism           3   
9    1/6/17               Close     random      terrorism           5   

   Injured  Total victims  Policeman Killed    Race Gender   Latitude  \
0       20             46               0.0   White      M        NaN   
1        0              3               0.0   White      M        NaN   
2        3              6               0.0   Black      M        NaN   
3      527            585               1.0   White      M  36.181271   
4        2              5               0.0   Asian      M        NaN   
5        0              3               NaN   White      M        NaN   
6        0              5               NaN     NaN      M        NaN   
7        0              3               1.0   White      M        NaN   
8        0              3               NaN   Black      M        NaN   
9        6             11               NaN  Latino      M        NaN   

    Longitude State  
0         NaN    TX  
1         NaN    CO  
2         NaN    MD  
3 -115.134132    NV  
4         NaN    CA  
5         NaN    PA  
6         NaN    FL  
7         NaN    OH  
8         NaN    CA  
9         NaN    FL

In [34]:
df1['State'].isna().sum()

45

In [35]:
df1.Cause.unique()

array(['unknown', nan, 'terrorism', 'unemployement', 'racism',
       'frustration', 'domestic dispute', 'anger', 'psycho', 'revenge',
       'domestic disputer', 'suspension', 'religious radicalism', 'drunk',
       'failing exams', 'breakup', 'robbery'], dtype=object)

We simplified the race into six categories

White, African American, Asian, Latino, Native America, and Others

In [36]:
##White
df1.replace({'Race':['White American or European American','white','White','White American or European American/Some other Race']},'White',inplace=True)
##Black 
df1.replace({'Race':['Black American or African American','Black','black','Black American or African American/Unknown']},'African American',inplace=True)
##Asian
df1.replace({'Race':['Asian American','Asian','Asian American/Some other race']},'Asian',inplace=True)
##Other
df1.replace({'Race':['Unknown','Some other race','Other','Two or more races']},'Other',inplace=True)


In [37]:
df1.Race.value_counts()

White                               144
African American                     85
Other                                66
Asian                                18
Latino                                5
Native American or Alaska Native      3
Name: Race, dtype: int64

In [38]:
df1.head()

Date Open/Close Location     Target    Cause  Fatalities  Injured  \
0   11/5/17               Close     random  unknown          26       20   
1   11/1/17                Open     random  unknown           3        0   
2  10/18/17               Close  coworkers  unknown           3        3   
3   10/1/17                Open     random  unknown          59      527   
4   6/14/17               Close  coworkers      NaN           3        2   

   Total victims  Policeman Killed              Race Gender   Latitude  \
0             46               0.0             White      M        NaN   
1              3               0.0             White      M        NaN   
2              6               0.0  African American      M        NaN   
3            585               1.0             White      M  36.181271   
4              5               0.0             Asian      M        NaN   

    Longitude State  
0         NaN    TX  
1         NaN    CO  
2         NaN    MD  
3 -115.134132    NV  
4         NaN    CA

We decided to leave NAs on some columns because it does not affect the drawing of our visualization. Also, since our dataset is small, it will affect our dataset dramatically. 

In [39]:
df1.isna().sum()

Date                    0
Open/Close Location    28
Target                  5
Cause                  77
Fatalities              0
Injured                 0
Total victims           0
Policeman Killed        6
Race                    2
Gender                  0
Latitude               20
Longitude              20
State                  45
dtype: int64

##### Ideas that we can explore with the mass shooting dataset

- Location of the event, along with numbers of victims that are injured and deceased
- Number of victims that are injured and deceased over time
- How many Policeman was killed
- Count of the race and gender of the shooter over time
